src/Symfony/Component/HttpKernel/RealHttpKernel.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpClient\HttpClient;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Mime\Part\AbstractPart;
+use Symfony\Component\Mime\Part\DataPart;
+use Symfony\Component\Mime\Part\Multipart\FormDataPart;
+use Symfony\Component\Mime\Part\TextPart;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * An implementation of a Symfony HTTP kernel using a "real" HTTP client.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+final class RealHttpKernel implements HttpKernelInterface
+{
+    private $client;
+    private $logger;
+
+    public function __construct(HttpClientInterface $client = null, LoggerInterface $logger = null)
```

Comment:
What about an event dispatcher to dispatch kernel events? Listeners could change the `http_client_options` attribute to prepare the request, but also do something with the response or exception.

Reply:
The idea could apply to all http clients actually.
Why not, in another PR :)